File: src/java/com/codecool/gameoflife/controller/GameController.java
package com.codecool.gameoflife.controller;

import com.codecool.gameoflife.model.Cell;
import com.codecool.gameoflife.model.GameBoard;
import com.codecool.gameoflife.model.RuleSet;
import com.codecool.gameoflife.view.GameView;

public class GameController {

    private GameBoard gameBoard;
    private GameView gameView;

    public void setGameBoard(GameBoard gameBoard) {
        this.gameBoard = gameBoard;
    }

    public void setGameView(GameView gameView) {
        this.gameView = gameView;
        this.gameView.updateView(gameBoard.getBoard());
    }

    public GameView getGameView() {
        return gameView;
    }

    public void nextGeneration() {
        this.gameBoard.nextGeneration();
        this.gameView.updateView(this.gameBoard.getBoard());
        }

    public void randomizeGrid() {
        this.gameBoard.randomizeBoard();
        this.gameView.updateView(this.gameBoard.getBoard());
    }

    public void updateCell(int x, int y) {
        Cell cellToUpdate = this.gameBoard.getBoard()[x][y];
        cellToUpdate.toggleAlive();
        this.gameView.updateCell(cellToUpdate, x, y);
    }

    public void toggleBorderlessMode() {
        this.gameBoard.toggleBorderLess();
    }

    public Cell[][] getGameBoard() {
        return this.gameBoard.getBoard();
    }

    public void setRules(RuleSet ruleSet) {
        this.gameBoard.setRuleSet(ruleSet);
    }

    public void clearBoard() {
        this.gameBoard.populateTiles();
        this.gameView.updateView(this.gameBoard.getBoard());
    }
}

File: src/kotlin/com/codecool/gameoflife/controller/GameController.kt
package com.codecool.gameoflife.controller

import com.codecool.gameoflife.model.GameBoard
import com.codecool.gameoflife.model.RuleSet
import com.codecool.gameoflife.view.GameView

class GameController(public val gameBoard: GameBoard) {
    val gameView: GameView

    init {
        gameView = GameView(this)
        gameView.updateView(gameBoard.board)
    }

    fun nextGeneration() {
        gameBoard.nextGeneration()
        gameView.updateView(gameBoard.board)
    }

    fun randomizeGrid() {
        gameBoard.randomizeBoard()
        gameView.updateView(gameBoard.board)
    }

    fun updateCell(x: Int, y: Int) {
        val cellToUpdate = gameBoard.board[x][y]
        cellToUpdate.toggleAlive()
        gameView.updateCell(x, y)
    }

    fun toggleBorderlessMode() {
        gameBoard.toggleBorderLess()
    }

    fun setRules(ruleSet: RuleSet) {
        gameBoard.setRuleSet(ruleSet)
    }

    fun clearBoard() {
        gameBoard.reset()
        gameView.updateView(gameBoard.board)
    }

    fun addRules(name: String, aliveString: String, deadString: String): Boolean {
        val aliveRules = convertRulesToList(aliveString)
        val deadRules = convertRulesToList(deadString)
        return if (deadRules.isEmpty() || aliveRules.isEmpty()) {
            false
        } else {
            RuleSet.addRuleSet(name, aliveRules, deadRules)
            gameView.updateRules()
            true
        }
    }

    private fun convertRulesToList(rules: String): List<Int> {
        return rules.split(",")
                .map { it.trim() }
                .filter { it.matches(RX_DIGITS) }
                .mapNotNull { it.toIntOrNull() }
    }

    companion object {
        private val RX_DIGITS = "\\d+".toRegex()
    }
}
